### PR TITLE
Sending all events from authed users through our servers (instead of directly to mixpanel)

### PR DESCRIPTION
**Pull request by stkem**
This is needed in order to augment the events with campaign information for the user and should make our lives a lot easier in the future as all (authed) events now go through one place (pending a change on iOS).

One commit in this PR per component changed (backend, angular site, extension)

Please review @andrewconner @2is10 @leogrim @yipingliao

### CODE REVIEW

**Comment by 2is10** [comment]
Mixpanel knows about users. Isn’t there a way to tell mixpanel a user’s campaign information just once?

https://mixpanel.com/docs/managing-users/managing-user_specific-properties
https://mixpanel.com/help/reference/javascript#storing-user-profiles

Would be sad to have to repeat it in every event.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/tracking/EventProxyController.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    actionAuthenticator: ActionAuthenticator,
+    heimdal: HeimdalServiceClient,
+    heimdalContextBuilderFactoryBean: HeimdalContextBuilderFactory) extends WebsiteController(actionAuthenticator) with ShoeboxServiceController {
+
+  private def jsValue2SimpleContextData(datum: JsValue): SimpleContextData = {
+    datum match {
+      case JsNumber(x) => ContextDoubleData(x.toDouble)
+      case JsBoolean(b) => ContextBoolean(b)
+      case JsString(s) => ContextStringData(s)
+      case huh => ContextStringData(huh.toString) //unexpected, trying our best not to loose data
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/tracking/EventProxyController.scala:23]: JsArray?

**Comment by stkem** [comment]
Unfortunately you can't filter events easily by user properties when doing segmentation/funnels :-(

File: kifi-backend/modules/shoebox/conf/shoebox.routes [hunk 2]
```diff
@@ -59,6 +59,14 @@ GET     /admin/unsubscribe          @com.keepit.controllers.website.EmailOptOutC
 GET     /extmsg/email/mute          @com.keepit.controllers.website.EmailNonUserMuteController.optOut(publicId: String)
 POST    /extmsg/email/mute          @com.keepit.controllers.website.EmailNonUserMuteController.optOutAction(publicId: String)
 
+
+##########################################
+# Tracking
+##########################################
+
+POST   /ev                          @com.keepit.controllers.tracking.EventProxyController.track()
```
Comment (2is10) [inline kifi-backend/modules/shoebox/conf/shoebox.routes]: As I mentioned in person, let’s use origin-specific routes, like `/ext/events` for the extension, for flexibility, as with every other API endpoint. They can all point to the same controller for now.

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/tracking/EventProxyController.scala (hunk 1 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/tracking/EventProxyController.scala:23]: Top level arrays are dealt with in the function that calls this. Nested arrays are not supported.

File: packrat/main.js [hunk 3]
```diff
@@ -213,15 +213,13 @@ function onGetFail(uri, done, failures, req) {
...
+var tracker = {
   enabled: true,
   queue: [],
   batch: [],
   sendBatch: function () {
     if (this.batch.length > 0) {
-      var json = JSON.stringify(this.batch);
-      var dataString = 'data=' + api.util.btoa(unescape(encodeURIComponent(json)));
-      api.postRawAsForm("https://api.mixpanel.com/track/", dataString);
+      ajax('POST', '/ev', this.batch.slice(0));
```
Comment (2is10) [inline packrat/main.js]: no need to `.slice`, `ajax` stringifies synchronously

File: packrat/main.js [hunk 4]
```diff
@@ -213,15 +213,13 @@ function onGetFail(uri, done, failures, req) {
...
-var mixpanel = {
+var tracker = {
   enabled: true,
   queue: [],
   batch: [],
   sendBatch: function () {
     if (this.batch.length > 0) {
-      var json = JSON.stringify(this.batch);
-      var dataString = 'data=' + api.util.btoa(unescape(encodeURIComponent(json)));
-      api.postRawAsForm("https://api.mixpanel.com/track/", dataString);
```
Comment (2is10) [inline packrat/main.js:224]: the `api.postRawAsForm` implementations can be removed too

File: kifi-backend/modules/shoebox/conf/shoebox.routes (hunk 2 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/conf/shoebox.routes]: done

File: packrat/main.js (hunk 3 above)
Comment (stkem) [inline packrat/main.js]: ah, good to know. removing.

File: kifi-backend/modules/shoebox/angular/src/common/services/mixpanelService.js [hunk 5]
```diff
@@ -38,6 +40,13 @@
     helpRankReKeeps: /^\/helprank\/rekeeps?$/
   };
 
+  function trackEventThroughProxy(event, properties)  {
+    return $http.post(env.origin + '/ev', [{
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/common/services/mixpanelService.js]: You shouldn’t need to specify `env.origin`. The browser resolves relative ajax URIs automatically.

**Comment by 2is10** [comment]
Minor comments. Otherwise looks good.

**Pull request merged**